### ISSUE TITLE
MainWindow for ipywidgets

### ISSUE THREAD

**Issue by tlambert03** [pyapp-kit/magicgui#601]
the Qt backend has a MainWindow but the ipywidgets backend doesn't.  As @larsoner has pointed out, the closest ipywidgets analog is probably [`AppLayout`](https://ipywidgets.readthedocs.io/en/7.x/examples/Layout%20Templates.html#AppLayout)

In Qt, a [main window](https://doc.qt.io/qt-6/qmainwindow.html#qt-main-window-framework) is characterized by the ability to add a menu bar (top), status bar (bottom), dock widgets, and toolbars:

![mainwindowlayout](https://github.com/pyapp-kit/magicgui/assets/1609449/5842e98a-7dad-48b5-9f8a-a8e7b935eea5)

In ipywidgets, it's more a layout issue, with a header, a footer, two sidebars and a central pane:

<img width="452" alt="Screen Shot 2023-10-11 at 5 24 57 PM" src="https://github.com/pyapp-kit/magicgui/assets/1609449/d3389ea9-1075-4153-a2c4-60d6c348d170">

Currently, our MainWindow protocol is a minimal subclass of `Container` that can add menus:

```python
class MainWindowProtocol(ContainerProtocol, Protocol):
    def _mgui_create_menu_item(
        self,
        menu_name: str,
        action_name: str,
        callback: Callable | None = None,
        shortcut: str | None = None,
    ) -> None:
```

but will likely gain `_mgui_add_toolbar` after #597...

This raises the question of how represent these somewhat different concepts of "main window" in our API.  If we use `AppLayout` behind the scenes, how do we assign stuff to left/middle/right?  Or do we just ignore the left and right (and allow people to manually edit using `widget.native` if they choose.  

fwiw, it looks like mne-python [doesn't use AppLayout](https://github.com/mne-tools/mne-python/blob/main/mne/viz/backends/_notebook.py) directly anywhere?

@larsoner, any insights/opinions from your mne applications?

**Comment by tlambert03** [comment]
thanks so much @larsoner ... that proposal seems as good as any to me!

> it might make the most sense to roughly follow the Qt naming scheme / model and squeeze this stuff into ipywidgets somehow

I agree with this too and i like your proposed model.  I'm sure we'll encounter some challenges with it at some point (related to mismatched user expectations)... but since the two models don't map exactly onto each other it does seem like we just have to make an opinion.

> Another option beyond AppLayout could be [ipywidgets GridSpecLayout](https://ipywidgets.readthedocs.io/en/7.x/examples/Layout%20Templates.html#Grid-layout). TBD if that's easier to get to work as a MainWindow-like "thing". I can play around with it a bit if it would help @tlambert03 !

Your help on that would be awesome.  I like what you've started with.  I'd be curious to hear what you would propose for methods and their behavior on the `magicgui.widgets.MainWindow` protocol.  that is, how exactly does a magicgui gui user put something in the each of the corresponding places?  would you go for `add_dock_widget`,  `add_toolbar `, `set_menu_bar`, `set_status_bar` like methods?  And those methods must be passed the corresponding `widget.ToolBar`, `widget.MenuBar`, etc... just like the Qt API?

**Comment by larsoner** [comment]
> would you go for add_dock_widget, add_toolbar , set_menu_bar, set_status_bar like methods? And those methods must be passed the corresponding widget.ToolBar, widget.MenuBar, etc... just like the Qt API?

Yeah that sounds good! `magicgui.widgets.StatusBar` can be [QStatusBar](https://doc.qt.io/qt-6/qstatusbar.html) in Qt and just a Container with a single Label or so in ipywidgets. Similar for the `ToolBar` I think. TBD how `MenuBar` could be implemented -- maybe that would need to be at the ipywidgets end first (or a [`DropDown`](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20List.html#dropdown) with some clever CSS and `on_change` event handling)? I think having ToolBar, StatusBar, and MenuBar classes should allow `magicgui` to adapt the behavior of these classes so they do what users expect. For example, adding a ToolBar to the top area should have the icons/buttons arranged horizontally, whereas adding it to the left area they should be arranged vertically, etc.

To keep things simple to start we could allow just a single widget per area: one ToolBar per toolbar area top/bottom/left/right (Qt allows multiple per area I think), a single Widget per dock area top/bottom/left/right, a single Widget as the central widget, etc. Someday if we want to allow multiple per area we can add a `append=False` or `replace=True` default kwarg or something but hopefully people can just use Containers if they need extra stuff.

**Comment by tlambert03** [comment]
Love it!
